### PR TITLE
Improve the signature display when a tuple appears many times within it.

### LINKED ISSUE #29949: VS Intellisense, generic parameters, and value-tuples
Hello Roslyn friends, I was asked to write an issue here since it was identified as something best addressed through Roslyn. However I originally phrased it as a VS issue, so my language reflects that somewhat, since I am not entirely certain how Roslyn is involved. So, please bear with me. 😄 
 
Visual Studio intellisense, hoverovers, and such things like this currently do not handle value-tuples in what I would consider a good way. However, other vaguely similar situations on more established idioms already have good practice, just maybe the UI hasn't quite caught up to value-tuples yet since they are so new.

The problem becomes obvious in cases where there is a library with generic type parameters, and value-tuples are ever used in those generic type parameters. LINQ would be the most obvious example, but I'll use immutable collections here. (No particular reason, I just happened to be playing with them in a scratch project, so that's what I wound up using here.)

```csharp 
using System;
using System.Collections.Generic;
using System.Collections.Immutable;
using System.Linq;
 
namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            var foo = ImmutableArray.Create(1, 2, 3, 4);
            var a = new { well = 5, gee = 3, that = 5, sounds = 2, terrific = 0 };
            var bar = ImmutableArray.Create(a, a, a, a);
            var t = (well: 5, gee: 3, that: 5, sounds: 2, terrific: 0);
            var biz = ImmutableArray.Create(t, t, t, t);
        }
    }
}
```

I'll give a series of examples, marked with ✔️ and ❌ examples to show what I feel are good and bad. I also list the good examples since I feel like they inform some ways in which we could address the bad examples.

1. ✔️ Cursor over the first lines `Create`:

   image

2. ✔️ Cursor over third line's `var`. This one is slightly odd, but at least it is comprehensible by a human.

   image

3. ✔️ Cursor over third line's `Create`:

   image

4. ✔️ Cursor over fifth line's `var`:

   image

5. ❌ Cursor over fifth line's `Create`:

   image

   This sort of seems like a missed opportunity. I would argue this is not easily comprehensible by a human, but it is easy to see how we might make it so. From example 3, I know these method *can* have sort of auxiliary explanatory information on the types of methods, and from example 4 I know that at least the type description is capable of calling them out.

   Hypothetically, in this case (generic parameter that is a value-tuple) maybe something that looks more like the following, in the case where the generic type parameter relevant to a method is a value-tuple, it could look like this. Please forgive my limited formatting skills:

   image

6. ✔️ Speaking not just of hover-overs but Intellisense, I think it's more or less the same but just to emphasize, if I cursor over the third line just prior to the `;`, then type `.`, we get this method `Add` as the first one. This looks pretty good.

   image

7. ❌  If I do the same on the fifth line, this is what `Add`'s intellisense display looks like:
   image

   Similarly to point 5, I feel like the enumeration of the type every single time it appears decreases the helpfulness of the display. We could imagine solving it in a similar fashion.

   image

/cc @jinujoseph @heejaechang @KrzysztofCwalina @zruty0

### LINKED ISSUE #33609: Signature info gets incomprehensible in presence of extension methods, generics and tuples
The displayed signature of `SelectAsArray` extension method is:

```C#
ImmutableArray<TResult> ImmutableArray<(DocumentId DocumentId, ImmutableArray<LineChange> Deltas)>.
SelectAsArray<(DocumentId DocumentId, ImmutableArray<LineChange> Deltas), TResult, TArg>(
Func<(DocumentId DocumentId, Immutable<ArrayChange> Deltas), TArg, TResult> map, TArg arg)
```

image

It's really hard to parse this when choosing between overloads of the method. We should think about better visualization of complex signatures.

The method itself is conceptually simple - it takes a lamda and applies it to elements of immutable array, but this is completely lost in the signature info. The problem is most apparent when tuples are substituted to generic parameters of Funcs.

### LINKED ISSUE #53061: Improve sig info, error message formatting when working with complex generic code
Working with types that combine generics and tuples is cumbersome. The IDE does not help much at all.

```C#
using var _ = PooledDictionary<string, ArrayBuilder<(ManagedActiveStatementDebugInfo info, SourceFileSpan span, int ordinal)>>.GetInstance(out var updatedSpansByDocumentPath);

var x = updatedSpansByDocumentPath.ToImmutableDictionary(
     keySelector: 
```

Can we structure the signatures better using line breaks and indentation?

Sig info:
image

Hover:
image

Types in diagnostic not colorized and the error is hard to read with the long type names in the middle of the error message:
image

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes #29949
Fixes #53061
Fixes #33609

Similar to what we do with anonymous types, we'll now show tuples like so (when they appear at least two times in a signature).

Before:

image

After:

image

### CODE REVIEW

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs [hunk 1]
```diff
@@ -517,6 +517,12 @@ private void AddTupleTypeName(INamedTypeSymbol symbol)
         {
             Debug.Assert(symbol.IsTupleType);
 
+            if (this.format.MiscellaneousOptions.IncludesOption(SymbolDisplayMiscellaneousOptions.DoNotExpandTupleType))
+            {
+                builder.Add(CreatePart(SymbolDisplayPartKind.StructName, symbol, "<tuple>"));
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs:522]: this is similar to what we do with anonymous types.  With anon-types, the compiler just makes a single part, allowing the IDE control of how to build the final display string.  For tuples it's fine (by default) to expand it as `(...)`, but we need a way to have that not happen in order to use the new display logic.

File: src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs [hunk 2]
```diff
@@ -534,7 +540,7 @@ private void AddTupleTypeName(INamedTypeSymbol symbol)
                 if (element.IsExplicitlyNamedTupleElement)
                 {
                     AddSpace();
-                    builder.Add(CreatePart(SymbolDisplayPartKind.FieldName, symbol, element.Name));
+                    builder.Add(CreatePart(SymbolDisplayPartKind.FieldName, element, element.Name));
```
Comment (CyrusNajmabadi) [inline src/Compilers/CSharp/Portable/SymbolDisplay/SymbolDisplayVisitor.Types.cs:543]: this was a bug before (we were adding in the tuple again for each tuple element).  This caused problems with the IDE code as we'd then try to expand these references out as well.

File: src/Compilers/Core/Portable/SymbolDisplay/SymbolDisplayMiscellaneousOptions.cs [hunk 3]
```diff
@@ -72,5 +72,11 @@ public enum SymbolDisplayMiscellaneousOptions
         /// Append '!' to non-nullable reference types.
         /// </summary>
         IncludeNotNullableReferenceTypeModifier = 1 << 8,
+
+        /// <summary>
+        /// Insert a tuple into the display parts as a single part instead of multiple parts (similar
+        /// to how anonymous types are inserted0.
+        /// </summary>
+        DoNotExpandTupleType = 1 << 9,
```
Comment (CyrusNajmabadi) [inline src/Compilers/Core/Portable/SymbolDisplay/SymbolDisplayMiscellaneousOptions.cs:80]: open to discussing a good name here.  We use the term 'expand' in other places (like 'ExpandNullable').  Maybe DoNotExpandTuples?

File: src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs [hunk 4]
```diff
@@ -7628,5 +7628,68 @@ void M()
...
+        [Fact, Trait(Traits.Feature, Traits.Features.QuickInfo)]
+        public async Task TestSingleTupleType()
+        {
+            await TestInClassAsync(
+@"void M((int x, string y) t) { }
+  void N()
+  {
+    $$M(default);
+  }",
+                MainDescription(@"void C.M((int x, string y) t)"),
```
Comment (CyrusNajmabadi) [inline src/EditorFeatures/CSharpTest/QuickInfo/SemanticQuickInfoSourceTests.cs:7641]: in the single use case, we keep the signature as before.

File: src/Features/CSharp/Portable/LanguageServices/CSharpStructuralTypeDisplayService.cs [hunk 5]
```diff
@@ -18,17 +18,17 @@
 
 namespace Microsoft.CodeAnalysis.Editor.CSharp.LanguageServices
 {
-    [ExportLanguageService(typeof(IAnonymousTypeDisplayService), LanguageNames.CSharp), Shared]
-    internal class CSharpAnonymousTypeDisplayService : AbstractAnonymousTypeDisplayService
+    [ExportLanguageService(typeof(IStructuralTypeDisplayService), LanguageNames.CSharp), Shared]
+    internal class CSharpStructuralTypeDisplayService : AbstractStructuralTypeDisplayService
```
Comment (CyrusNajmabadi) [inline src/Features/CSharp/Portable/LanguageServices/CSharpStructuralTypeDisplayService.cs:22]: most of the code changes are a strictly mechanical change of 'AnonymousType' => 'StructuralType'.  I'll mark which parts of the code are actually changed.

File: src/Features/Core/Portable/LanguageServices/AnonymousTypeDisplayService/AbstractAnonymousTypeDisplayService.cs [hunk 6]
```diff
@@ -1,176 +0,0 @@
-﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/LanguageServices/AnonymousTypeDisplayService/AbstractAnonymousTypeDisplayService.cs:1]: this type was not deleted.  but enough changed happened to it that github considers it a rewrite.

File: src/Features/Core/Portable/LanguageServices/AnonymousTypeDisplayService/AbstractStructuralTypeDisplayService.StructuralTypeCollectorVisitor.cs [hunk 7]
```diff
@@ -75,27 +69,37 @@ public override void VisitModule(IModuleSymbol symbol)
 
             public override void VisitNamedType(INamedTypeSymbol symbol)
             {
+                if (_namedTypes.TryGetValue(symbol, out var orderAndCount))
+                {
+                    orderAndCount.count++;
+                    _namedTypes[symbol] = orderAndCount;
+                    return;
+                }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/LanguageServices/AnonymousTypeDisplayService/AbstractStructuralTypeDisplayService.StructuralTypeCollectorVisitor.cs:77]: this is new logic.

File: src/Features/Core/Portable/LanguageServices/AnonymousTypeDisplayService/AbstractStructuralTypeDisplayService.cs [hunk 8]
```diff
@@ -0,0 +1,198 @@
...
+                typeParts.AddRange(Space());
+
+                if (structuralType.IsValueType)
+                {
+                    typeParts.AddRange(structuralType.ToMinimalDisplayParts(semanticModel, position));
+                }
+                else
+                {
+                    typeParts.AddRange(GetTypeParts(structuralType, semanticModel, position));
+                }
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/LanguageServices/AnonymousTypeDisplayService/AbstractStructuralTypeDisplayService.cs:61]: this part is new.

File: src/Features/Core/Portable/LanguageServices/AnonymousTypeDisplayService/AbstractStructuralTypeDisplayService.cs [hunk 9]
```diff
@@ -0,0 +1,198 @@
...
+                        }
+                    }).ToImmutableArray();
+            }
+
+            return structuralTypes;
+        }
+
+        private static ImmutableArray<INamedTypeSymbol> GetTransitiveStructuralTypeReferences(
+            ImmutableArray<INamedTypeSymbol> structuralTypes)
+        {
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/LanguageServices/AnonymousTypeDisplayService/AbstractStructuralTypeDisplayService.cs:142]: this method has been rewritten.

File: src/Features/Core/Portable/LanguageServices/AnonymousTypeDisplayService/StructuralTypeDisplayInfo.cs [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+        public IList<SymbolDisplayPart> ReplaceStructuralTypes(IList<SymbolDisplayPart> parts, SemanticModel semanticModel, int position)
+            => ReplaceStructuralTypes(parts, StructuralTypeToName, semanticModel, position);
+
+        public static IList<SymbolDisplayPart> ReplaceStructuralTypes(
+            IList<SymbolDisplayPart> parts,
+            IDictionary<INamedTypeSymbol, string> structuralTypeToName,
+            SemanticModel semanticModel,
+            int position)
+        {
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/LanguageServices/AnonymousTypeDisplayService/StructuralTypeDisplayInfo.cs:31]: this method was rewritten.

File: src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AnonymousTypes.cs [hunk 11]
```diff
@@ -39,32 +39,23 @@ private void InlineAllDelegateAnonymousTypes()
...
+                    firstSymbol, directStructuralTypes.ToImmutableArrayOrEmpty(), _semanticModel, _position);
 
-                if (info.AnonymousTypesParts.Count > 0)
+                if (info.TypesParts.Count > 0)
                 {
-                    AddToGroup(SymbolDescriptionGroups.AnonymousTypes,
-                        info.AnonymousTypesParts);
+                    AddToGroup(SymbolDescriptionGroups.StructuralTypes, info.TypesParts);
 
-restart:
```
Comment (CyrusNajmabadi) [inline src/Features/Core/Portable/LanguageServices/SymbolDisplayService/AbstractSymbolDisplayService.AnonymousTypes.cs:58]: this loop moved into StructuralTypeDisplayInfo.ReplaceStructuralTypes

**Comment by davidwengier** [comment]
I was expecting this to be a new section on the tool tip. Did it get too long with a new section, or do you just consider them the same? And if so, how much is that your compiler-dev brain at work? :)

"Structural types" doesn't have much meaning to me, though I'm not exactly against it, as it I can work out what it means. Mainly just surprised, especially if adding delegate info is going to give us a new section here anyway.

File: src/Features/Core/Portable/LanguageServices/AnonymousTypeDisplayService/AbstractStructuralTypeDisplayService.cs [hunk 12]
```diff
@@ -122,24 +134,36 @@ private static string GenerateAnonymousTypeName(int current)
...
+            foreach (var type in structuralTypes)
                 type.Accept(visitor);
+
+            using var _ = ArrayBuilder<INamedTypeSymbol>.GetInstance(out var result);
+
+            foreach (var (namedType, (order, count)) in transitiveReferences.OrderBy(kvp => kvp.Value.order))
+            {
+                if (namedType.IsTupleType && count == 1)
+                {
+                    // Ignore tuples only referenced once.  We'll keep them inline.  If the tuple shows up multiple times,
```
Comment (davidwengier) [inline src/Features/Core/Portable/LanguageServices/AnonymousTypeDisplayService/AbstractStructuralTypeDisplayService.cs:158]: I wonder if this should only apply if there isn't otherwise going to be a "Structural Types" section? eg, what happens in this case:

```
public (int a, int b) Goo((int a, int b), (string x, string y);
```

You're going to end up with a "Structural Types" section that doesn't list out all of the structural types.

**Review by davidwengier — APPROVED** [review]
Another random thought, is it worth using `'t` for tuples to distinguish them?

**Comment by CyrusNajmabadi** [comment]
> Did it get too long with a new section

I just didn't think this warranted a new section. To me each section is very distinct in it's purpose. If we added a new section it would effectively be done the exact same as the existing anonymous type section. 

Because of that, I just washed one section that showed you all the type expansions in a single place.

**Comment by CyrusNajmabadi** [comment]
> especially if adding delegate info is going to give us a new section here anyway.

I would definitely put delegates into this section, and I had that in mind when writing this :-)

**Comment by CyrusNajmabadi** [comment]
> Structural types" doesn't have much meaning to me, though I'm not exactly against it,

You're right. I think that can just simplify to `Types:` and remain just as clear.

File: src/Features/Core/Portable/LanguageServices/AnonymousTypeDisplayService/AbstractStructuralTypeDisplayService.cs (hunk 12 above)
Comment (CyrusNajmabadi) [reply src/Features/Core/Portable/LanguageServices/AnonymousTypeDisplayService/AbstractStructuralTypeDisplayService.cs:158]: That's an interesting point. I could revise to make it that once we have at least one duplicated tuple that all tuples then are broken out...  That seems reasonable to me.

**Comment by CyrusNajmabadi** [comment]
> Another random thought, is it worth using 't for tuples to distinguish them?

Possibly. Are you thinking `'t1, 't2, etc `?

**Comment by davidwengier** [comment]
> > Another random thought, is it worth using 't for tuples to distinguish them?
> 
> Possibly. Are you thinking `'t1, 't2, etc `?

Nevermind, ignore me. I always assumed the "a" in `'a` stood for "anonymous".. just tried with two and it turns out its just alphabetical.. this is fine :)

**Comment by 333fred** [comment]
@CyrusNajmabadi I don't think this requires a full session, but can you please send an email to the apu design alias with the proposal?

**Comment by CyrusNajmabadi** [comment]
On it.

**Comment by akhera99** [comment]
Fixes: #29949 #53061 #33609

**Comment by CyrusNajmabadi** [comment]
@333fred renamed to CollapseTupleTypes

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler  for small public API change to enhance SymbolDisplay.

**Comment by CyrusNajmabadi** [comment]
@dotnet/roslyn-compiler @333fred for eyes on the SymbolDisplay change.

**Comment by CyrusNajmabadi** [comment]
Rebasing to 17.1

**Review by 333fred — APPROVED** [review]
Symbol display change LGTM

**Comment by CyrusNajmabadi** [comment]
@jcouv can you ptal?

File: src/Compilers/Core/Portable/SymbolDisplay/SymbolDisplayMiscellaneousOptions.cs [hunk 13]
```diff
@@ -72,5 +72,11 @@ public enum SymbolDisplayMiscellaneousOptions
         /// Append '!' to non-nullable reference types.
         /// </summary>
         IncludeNotNullableReferenceTypeModifier = 1 << 8,
+
+        /// <summary>
+        /// Insert a tuple into the display parts as a single part instead of multiple parts (similar
+        /// to how anonymous types are inserted).
+        /// </summary>
+        CollapseTupleTypes = 1 << 9,
```
Comment (cston) [inline src/Compilers/Core/Portable/SymbolDisplay/SymbolDisplayMiscellaneousOptions.cs:80]: >CollapseTupleTypes [](http://example.com/codeflow?start=8&length=18)

I'm curious why special-case tuples and anonymous types. Wouldn't similar behavior be useful for other type arguments?

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/SymbolDisplay/SymbolDisplayMiscellaneousOptions.cs:80]: Primarily because they are structural types that require a lot more than a name to reference. For example, when you have a nominal type, it doesn't matter how many members that type has, you just reference it by name. Hit with a structural type you have to include that structure in somehow. 

If we have more cases like that in the future, I'd expect we go down a similar route.

Comment (cston) [reply src/Compilers/Core/Portable/SymbolDisplay/SymbolDisplayMiscellaneousOptions.cs:80]: > when you have a nominal type, it doesn't matter how many members that type has, you just reference it by name

But the nominal type could be arbitrarily complicated, for instance `Dictionary<string, object>`.

Comment (CyrusNajmabadi) [reply src/Compilers/Core/Portable/SymbolDisplay/SymbolDisplayMiscellaneousOptions.cs:80]: Potentially. We haven't had that arise as significant enough to need attention. If we did though, I'd prefer that as a separate flag.

**Review by jcouv — APPROVED** [review]
Compiler change LGTM Thanks (iteration 34). I take Fred's sign-off as evidence of sufficient public API discussion.
